### PR TITLE
Closes #3280: Add isServiceAvailable to PushService 

### LINKED ISSUE #3280: Consider adding isServiceAvailable to PushService
Adding this method to the PushService will make the implementation across various applications easier to follow.

API:
```kotlin
interface PushService {
  fun isServiceAvailable(context: Context): Boolean
}
```

### PR DESCRIPTION
**Pull request by jonalmeida**
Added the implementations of `isServiceAvailable` for Firebase and Amazon. I spent a bit trying to better test the exception part for ADM without making the implementation more complicated to no avail.

Also fixed a bug where registration errors were not being forwarded to the processor.

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt [hunk 1]
```diff
@@ -74,6 +74,21 @@ abstract class AbstractAmazonPushService : ADMMessageHandlerBase("AbstractAmazon
...
 
+    override fun isServiceAvailable(context: Context): Boolean {
+        // This is a strange API: the ADM SDK may exist on the device if it's an Amazon device, but we
+        // also need to check if it's a *supported* Amazon device then.
+        // We create a temporary instance of the ADM class, if it's available so that we can call `isSupported`.
+        // This is the recommended approach as per the docs:
+        // https://developer.amazon.com/docs/adm/integrate-your-app.html#gracefully-degrade-if-adm-is-unavailable
+        return try {
+            Class.forName("com.amazon.device.messaging.ADM")
+            ADM(context).isSupported
```
Comment (Amejia481) [inline components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt]: Amazon is the new Samsung 🦀  XD

**Comment by jonalmeida** [comment]
Ah, I lied on my PR checklist. I didn't run ktlint on it. :(

> LGTM 👍
> 
> Let's fix a small ktlint issue before merging.
> 
> ```shell
> > Task :ktlint
> /build/android-components/components/lib/push-amazon/src/test/java/mozilla/components/lib/push/amazon/AbstractAmazonPushServiceTest.kt:75:29: Missing spacing before ":"
> 
> > Task :ktlint FAILED
> 
> FAILURE: Build failed with an exception.
> ```

File: components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt (hunk 1 above)
Comment (jonalmeida) [reply components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt]: Honestly!

**Comment by jonalmeida** [comment]
bors r+

After fixing the ktlint issue.

**Pull request merged**